File: leakcanary-support-fragment/src/androidTest/java/com/squareup/leakcanary/tests/Fragments.java
package com.squareup.leakcanary.tests;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import java.util.concurrent.CountDownLatch;

public final class Fragments {

  public static CountDownLatch waitForFragmentDetached(FragmentActivity activity) {
    final CountDownLatch latch = new CountDownLatch(1);
    final FragmentManager fragmentManager = activity.getSupportFragmentManager();
    fragmentManager.registerFragmentLifecycleCallbacks(
        new FragmentManager.FragmentLifecycleCallbacks() {
          @Override public void onFragmentDetached(FragmentManager fm, Fragment f) {
            fragmentManager.unregisterFragmentLifecycleCallbacks(this);
            latch.countDown();
          }
        }, false);
    return latch;
  }

  public static CountDownLatch waitForFragmentViewDestroyed(FragmentActivity activity) {
    final CountDownLatch latch = new CountDownLatch(1);
    final FragmentManager fragmentManager = activity.getSupportFragmentManager();
    fragmentManager.registerFragmentLifecycleCallbacks(
        new FragmentManager.FragmentLifecycleCallbacks() {
          @Override public void onFragmentViewDestroyed(FragmentManager fm, Fragment f) {
            fragmentManager.unregisterFragmentLifecycleCallbacks(this);
            latch.countDown();
          }
        }, false);
    return latch;
  }

  private Fragments() {
    throw new AssertionError();
  }
}

File: leakcanary-support-fragment/src/androidTest/java/com/squareup/leakcanary/tests/Fragments.kt
package com.squareup.leakcanary.tests

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import java.util.concurrent.CountDownLatch

class Fragments private constructor() {

  init {
    throw AssertionError()
  }

  companion object {
    fun waitForFragmentDetached(activity: FragmentActivity): CountDownLatch {
      val latch = CountDownLatch(1)
      val fragmentManager = activity.supportFragmentManager
      fragmentManager.registerFragmentLifecycleCallbacks(
          object : FragmentManager.FragmentLifecycleCallbacks() {
            override fun onFragmentDetached(
              fm: FragmentManager,
              f: Fragment
            ) {
              fragmentManager.unregisterFragmentLifecycleCallbacks(this)
              latch.countDown()
            }
          }, false
      )
      return latch
    }

    fun waitForFragmentViewDestroyed(activity: FragmentActivity): CountDownLatch {
      val latch = CountDownLatch(1)
      val fragmentManager = activity.supportFragmentManager
      fragmentManager.registerFragmentLifecycleCallbacks(
          object : FragmentManager.FragmentLifecycleCallbacks() {
            override fun onFragmentViewDestroyed(
              fm: FragmentManager,
              f: Fragment
            ) {
              fragmentManager.unregisterFragmentLifecycleCallbacks(this)
              latch.countDown()
            }
          }, false
      )
      return latch
    }
  }
}
